### ISSUE TITLE
Dask Parallelism

### ISSUE THREAD

**Issue by andersy005** [klindsay28/CESM2_coup_carb_cycle_JAMES#1]
@klindsay28, 
- If you are planning on running `gen_tseries.py` via a batch job, I would recommend using [`dask-mpi`](http://mpi.dask.org/en/latest/), which can be installed via:
  -  pip: `pip install dask_mpi --upgrade` or 
  -  conda: `conda install dask-mpi -c conda-forge`

Once `dask-mpi` is installed, you will need the following modifications:

```python
#! /usr/bin/env python

from dask_mpi import initialize
initialize()
from distributed import Client


import argparse
from datetime import datetime, timezone
import math
import os

import xarray as xr
import yaml
from cf_units import Unit

import data_catalog
from tseries_utils import clean_units, get_weight, tseries_fname

client = Client()
```

After requesting resources from the scheduler, you can run:

```console
mpirun -np 4 python gen_tseries.py
```

This will run the Dask Scheduler on MPI rank 0, the user’s Client code on MPI rank 1, and 2 workers on MPI rank 2 and MPI rank 3. So if you need 10 workers, you would run:

```console
mpirun -np 12 python gen_tseries.py
```

@matt-long suggested a different approach that uses `dask-jobqueue` for interactive jobs:

```python
from dask_jobqueue import PBSCluster
cluster = PBSCluster(queue='regular',
                     cores = 18,
                     processes = processes,
                     memory = '100GB',          
                     project = project,
                     walltime = '04:00:00',
                     local_directory=f'/glade/scratch/{USER}/dask-tmp')
```
Let me know me if this is what you would like to do instead.

**Comment by klindsay28** [comment]
@andersy005, I installed dask-mpi into my conda environment.

(FYI, it seems like one cannot do this at the command prompt in a jupyterhub terminal. I also couldn't run git push from a jupyterhub terminal. It looks like there are some restrictions on network access from within jupyterhub.)

I tried adding the lines above to a jupyter notebook, gen_tseries_testing.ipynb, and running it within jupyterhub, but I get an error with the `dask_mpi.initialize()` call:

> RuntimeError: Cannot run the event loop while another loop is running

Suggestions?

**Comment by klindsay28** [comment]
I had fired up jupyterhub with 4 tasks and 4 CPUs.

**Comment by andersy005** [comment]
> (FYI, it seems like one cannot do this at the command prompt in a jupyterhub terminal. I also couldn't run git push from a jupyterhub terminal. It looks like there are some restrictions on network access from within jupyterhub.)

For the time being, Jhub kernels are being launched on Cheyenne compute nodes (which don't have access to the internet).


> I had fired up jupyterhub with 4 tasks and 4 CPUs.

Because JHub kernel is running inside a regular batch job, this makes it very hard (if not impossible) to use `dask-mpi` in this environment. 

My suggestion would be to use a basic PBS script for running an MPI job:

`gen_tseries.sh`:

```bash
#!/bin/bash
### Job Name
#PBS -N dask_mpi_job
### Project code
#PBS -A project_code
#PBS -l walltime=01:00:00
#PBS -q queue_name
### Merge output and error files
#PBS -j oe
### Select 2 nodes with 36 CPUs each for a total of 72 MPI processes
#PBS -l select=2:ncpus=36:mpiprocs=36
### Send email on abort, begin and end
#PBS -m abe
### Specify mail recipient
#PBS -M email_address



### Run the executable
mpirun -np 72 python gen_tseries.py
```

```console
qsub gen_tseries.sh
```

**Comment by klindsay28** [comment]
@andersy005, is there a way to use dask-mpi within a notebook on cheyenne or casper?

Following your advice, I'm trying to do testing of my python code within a notebook prior to going full speed with a non-interactive script.

**Comment by andersy005** [comment]
> is there a way to use dask-mpi within a notebook on cheyenne or casper?

Not that I know of. Ccing @kmpaul

**Comment by klindsay28** [comment]
Thanks @andersy005. I'll give the `dask_jobqueue.PBSCluster` suggestion of @matt-long a try.

**Comment by kmpaul** [comment]
There is probably an event loop running on the Jupyter notebook and Dask MPI can’t run one on its own.

**Comment by klindsay28** [comment]
fixed with commit 52d18c43266747dcae94453bb1c1a45a2c6da084, which enables dask parallelism in `tseries_mod.tseries_gen` by using `dask_jobqueue.PBSCluster`
